File: W2_0021-merge-two-sorted-lists/merge-two-sorted-lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    const dummy = new ListNode(0);
    let cur = dummy;

    let next = getSmallerNode(list1, list2);
    while (next !== null) {
        cur.next = next;
        cur = cur.next;

        [list1, list2] = advanceLists(list1, list2, next);
        next = getSmallerNode(list1, list2);
    }

    return dummy.next;
}


function getSmallerNode(node1: ListNode | null, node2: ListNode | null): ListNode | null {
    if (node1 === null) {
        return node2;
    }
    if (node2 === null) {
        return node1;
    }
    return node1.val < node2.val ? node1 : node2;
}


function advanceLists(node1: ListNode | null, node2: ListNode | null, chosen: ListNode): [ListNode | null, ListNode | null] {
    if (chosen === node1) {
        return [node1?.next || null, node2];
    }
    return [node1, node2?.next || null];
}

File: W2_0021-merge-two-sorted-lists/merge-two-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	cur := dummy

	next := getSmallerNode(list1, list2)
	for next != nil {
		cur.Next = next
		cur = cur.Next

		list1, list2 = advanceLists(list1, list2, next)
		next = getSmallerNode(list1, list2)
	}

	return dummy.Next
}

func getSmallerNode(node1 *ListNode, node2 *ListNode) *ListNode {
	if node1 == nil {
		return node2
	}
	if node2 == nil {
		return node1
	}
	if node1.Val < node2.Val {
		return node1
	}
	return node2
}

func advanceLists(node1 *ListNode, node2 *ListNode, chosen *ListNode) (*ListNode, *ListNode) {
	if chosen == node1 {
		return node1.Next, node2
	}
	return node1, node2.Next
}